### DISCUSSION TITLE
Adicionar vários exemplos de código

### DISCUSSION THREAD

**Discussion by armindomatheus** [levxyca/diciotech#54] [Ideas]
Na na expressões ternárias temos a presença de um código como exemplo, porém em certos casos acho que podemos mostrar um exemplo em várias linguagens diferentes, já que dependo do exemplo a sintaxe pode mudar de linguagem pra linguagem

**Comment by freitaschz** [comment]
Creio que é uma boa sim, mas não sei qual seria a melhor forma de implementar isso. Temos que discutir aqui qual seria o ideal e quais linguagens devem estar presentes, pois não podem ser muitas e sim talvez umas quatro.

Posso fazer um PR de uma primeira versão dessa feature! 🫡 Claro, de acordo com o que for discutido nessa issue e se a @levxyca, junto com a comunidade, concordar na implementação dessa sugestão.

**Comment by levxyca** [comment]
Eu acho válido abrir a discussão para a comunidade, posso cirar um tópico novo nos `discussions` do repositório, meu único ponto quanto a isso é fugir muito da ideia de explicar os termos em si, o código de exemplo é mais um apoio para a pessoa conseguir visualizar, mas a ideia realmente não é ensinar implementar aqui, faz sentido?

@armindomatheus @thiagofqs

**Comment by freitaschz** [comment]
> Eu acho válido abrir a discussão para a comunidade, posso cirar um tópico novo nos `discussions` do repositório, meu único ponto quanto a isso é fugir muito da ideia de explicar os termos em si, o código de exemplo é mais um apoio para a pessoa conseguir visualizar, mas a ideia realmente não é ensinar implementar aqui, faz sentido?
> 
> @armindomatheus @thiagofqs

Concordo, @levxyca!

Bem, já deixando minha contribuição nessa discussão, penso da seguinte forma:
Mesmo o objetivo do Diciotech sendo explicar os termos em si, acho válido uma exemplificação com uma linguagem ou outra, mas muito poucas. Por exemplo, se vier algum conceito que é necessário ser exemplificado através de um código, será que é possível fazê-lo sempre no padrão como é no termo do operador ternário `Expressão ? Verdadeiro : Falso`? Pode ser que sim, mas também pode não ficar muito claro para alguma pessoa usuária que está em busca do conceito do termo.

**Comment by levxyca** [comment]
Uma sugestão, pensando em implementação, seria alinhar isso com a nova UI que está sendo desenvolvida na issue #12

**Comment by freitaschz** [comment]
> Uma sugestão, pensando em implementação, seria alinhar isso com a nova UI que está sendo desenvolvida na issue #12

Siimm! Se realmente formos implementar, antes temos que procurar padronizar um pouco as coisas, como incluir algumas regras e requisitos para incluir novos termos no arquivo JSON.

**Comment by levxyca** [comment]
> > Uma sugestão, pensando em implementação, seria alinhar isso com a nova UI que está sendo desenvolvida na issue #12
> 
> Siimm! Se realmente formos implementar, antes temos que procurar padronizar um pouco as coisas, como incluir algumas regras e requisitos para incluir novos termos no arquivo JSON.

Exatamente, vou converter essa issue em uma discussion para nos organizarmos melhor!

**Comment by freitaschz** [comment]
@levxyca, o que sugeres com relação a essa feature? Se for realmente válido, é interessante cita-la na issue #12 para, se possível, o @kvnol implementar na V2 da UI do Diciotech como já está presente na V1 da UI, do contrário pode ser utilizado o mesmo espaço para outra implementação.

**Reply by levxyca** [reply]
Acho que podemos coletar os feedbacks da comunidade e entender se faz sentido, pensei em criar uma outra enquete aqui e divulgar para entendermos isso melhor. O que acha @thiagofqs?

**Reply by freitaschz** [reply]
Com certeza, @levxyca! Creio que é o melhor a se fazer.

**Comment by rwietter** [comment]
O _layout_ proposto pela issue #12 é bem legal. Vou colocar outra sugestão, que seria usar o `código` como `pseudocódigo` para conceitos gerais, já que estamos tratando de conceituação, como abordado em muitos livros de programação, tal como em **"Algoritmos: Teoria e Prática" de Thomas H. Cormen**. O `pseudocódigo` é **agnóstico a linguagem** e trata de **conceituar de forma ampla**. Então colocar links de artigos e referências para ver em linguagens, tal como esse [repositório de algoritmos](https://www.rosettacode.org/wiki/Factorial).

## [Ternário](https://www.rosettacode.org/wiki/Conditional_structures)

O ternário geralmente é bastante similar em muitas linguagens como em Java, C#, PHP, JS, Swift, Ruby, a maioria usa `? :`. O Python troca as expressões, mas aparenta ser muito mais um `if-else` do que ternário em si, por exemplo:

```py
resultado = 1 if 1 == 1 else 0 # 1
```

Ternário em `pseudocódigo`

```
resultado <- condição ? valor_se_verdadeiro : valor_se_falso
```

ou

```
result <- booleanExpression ? expression1 : expression2
```

## [Fatorial](https://www.rosettacode.org/wiki/Factorial)

No caso de um **fatorial** em pseudocódigo mais **descritivo**

```
Entrada: Um inteiro n ≥ 0.
Saída: O resultado do fatorial de (n).

1. Se n = 0, então retorne 1 como saída.
2. Caso contrário, retorne n vezes o valor retornado chamando fatorial(n − 1) recursivamente.
```

O Thomas Cormen usa em "Desmistificando Algoritmos" uma metodologia bastante descritiva. Exemplo:

![Desmistificando Algoritmos](https://i.imgur.com/nGb72jA.png)

**Fatorial** mais `pseudocódigo`

```
function factorial(n)
    if n equals 0
        return 1
    else
        return n * factorial(n - 1)
    end if
end function
```

Já no livro "Algoritmos: Teoria e Prática", Cormen faz o `pseudocódigo` de forma mais sucinta e mais próximo do código real, mas é uma exemplificação agnóstica que pode ser entendida independente de qual linguagem de programação conheça. Exemplo:

![Algoritmos: Teoria e Prática](https://imgur.com/SgQ62KN.png)

A ideia de ter código para várias linguagens é interessante, mas pelo tamanho do card e issues e PRs que devem ser abertos para adicionar variações de linguagens torna-se bastante trabalhoso e em alguns casos até desnecessário.

**Comment by freitaschz** [comment]
Perfeita colocação, @rwietter! Concordo totalmente contigo.
Creio que códigos de exemplo, em linguagens específicas, devem ser apenas utilizados em conceitos específicos de uma linguagem ou quando realmente não der para exemplificar, de forma simples e objetiva, com pseudocódigo.

**Comment by levxyca** [comment]
Achei sensacional seu comentário @rwietter! Faz total sentido sim! Podemos adicionar isso no guia de contribuições e manter como uma boa prática na comunidade, acho que a solução mais didática e simples que podemos adotar!

**Comment by george-gca** [comment]
Eu acredito que geralmente o usuário vai se interessar mais pela implementação em uma linguagem específica. E opinião pessoal, mas acho que pseudo código as vezes é mais difícil de entender do que ver uma solução implementada em uma linguagem que se está aprendendo. Pelo menos quando eu aprendi a programar era muito mais fácil entender um exemplo em C (que foi a 1a linguagem que eu aprendi) do que em pseudo código. As vezes o pseudo código esconde algumas nuâncias que são específicas de linguagem que podem dificultar a pessoa que tá aprendendo a entender como transformar aqueles passos em um código real.

Eu pensei no seguinte: e se tivesse como o usuário de alguma forma selecionar as linguagens que ele tem interesse em ver? Nem que fosse ali em cima perto do botão de mudar o tema, tivesse um dropdown com todas as linguagens suportadas e um checkbox do lado do nome (incluindo pseudo código)? Aí quando fosse mostrar exemplos em código, só iria aparecer as opções de linguagens que ele selecionou (supondo que exista o exemplo). Sei que tem alguns repositórios que implementam [alguns algoritmos já em várias linguagens](https://github.com/thuva4/Algorithms) e até um só de [hello world](https://github.com/arjuncvinod/Hello-World-in-Different-Languages).

De repente até os exemplos de código poderiam ser um repositório a parte e esse só referenciar de alguma forma.

**Reply by levxyca** [reply]
Achei a ideia muito legal @george-gca! Realmente, acho que isso varia de pessoa pra pessoa, eu por exemplo, no ínicio, tinha mais facilidade de entender os conceitos com pseudo código do que com uma linguagem específica. Podemos amadurecer melhor essa ideia! Pensar em como podemos fazer essa seleção de linguagens e como mostrar isso para as pessoas usuárias 🤔

Acredito que mesmo nesse cenário, seja interessante manter o pseudo código e até mesmo deixar ele como default, tendo em visto que o Diciotech não é apenas para pessoas programadoras!

**Reply by george-gca** [reply]
Concordo. Mas pensando em pessoas não programadoras, não sei se o pseudo código seria útil pra elas também. Independente, ter opções é interessante.

**Reply by levxyca** [reply]
Legal! Eu gosto do pseudo código para pessoas não programadoras também, pois acredito que seja mais fácil de compreender do que uma linguagem de programação em si, mas concordo que ter opções é interessante.

Acho que agora, a ideia é entender melhor qual a melhor forma de aprensentar esses exemplos dentro do projeto. Por exemplo, eu penso em algumas questões, como definir quais linguagens colocar de exemplo?

**Reply by george-gca** [reply]
Acho que isso pode ser baseado nas linguagens de entrada. Por exemplo, muitos cursos de introdução a programação tanto em universidade quanto em outros cursos online são hoje dados em python. Acredito que alguns ainda sejam dados em C também. Então acho que essas 2 linguagens deveriam ser as prioridades.
